src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -285,10 +292,8 @@ private void MarkEntireTypeInternal (TypeDefinition type, bool includeBaseTypes,
 
 			if (type.HasMethods) {
 				foreach (MethodDefinition method in type.Methods) {
-					// Probably redundant since we EnqueueMethod below anyway.
-					Annotations.Mark (method, new DependencyInfo (DependencyKind.MemberOfType, type));
 					Annotations.SetAction (method, MethodAction.ForceParse);
-					EnqueueMethod (method, new DependencyInfo (DependencyKind.MemberOfType, type));
+					MarkMethod (method, new DependencyInfo (reason.Kind == DependencyKind.AccessedViaReflection ? reason.Kind : DependencyKind.MemberOfType, type), new MessageOrigin (reason.Source as IMemberDefinition));
```

Comment:
Why the special case for `AccessedViaReflection` here - is it just to ensure the reason is one of the expected ones in ProcessRequiresUnreferencedCode?
- Can we do this without changing the reason for the marked method?
- If we go with the special case, should we do the same for `DependencyKind.DynamicDependency` - since we can get here from `MarkDynamicDependency`?

Reply:
> Why the special case for AccessedViaReflection here - is it just to ensure the reason is one of the expected ones in ProcessRequiresUnreferencedCode?

Yes.

> Can we do this without changing the reason for the marked method?

We can do this if we add `DynamicDependency` to the list of reasons for marking a method... I think it makes sense and should make code cleaner/easier to read.

Reply:
I'm not sure you'd need to add `DynamicDependency` if we continue passing `MemberOfType` - I think you'd need to add `MemberOfType` to the list of cases in `ProcessRequiresUnreferencedCode`.

Reply:
Just tried not special casing. If we simple reuse `reason.Kind` then we can end up passing `BaseType` or `NestedType`, which then throws, since these are not valid reasons for marking a method.
The reason to special case here is so that we warn on annotated methods on a type that is kept because of `DynamicallyAccessedMembers.All`. We don't have `MembersOfType` on the list of reasons to process `ProcessRequiresUnreferencedCode` since we don't want to emit warnings on methods that are marked because of a copied assembly (we only want to warn if there's a possibility of the method being referenced through reflection or if it's directly called).
I think it's easier to just keep this special case and of course, special case `DynamicDependency`  too.

Reply:
Ah, that makes sense, to avoid warnings for copy assemblies. Thanks for the explanation!

Just another thought - now I am wondering if we could use sourceLocationMember only, instead of DependencyInfo, for this purpose - since DependencyInfo was designed to be more granular than source locations, and what we need here are source locations. But that might require more changes to the sourceLocationMember logic - I'm not sure. I'll leave it as a suggestion in case you think it's worth looking into, but I think special-casing `AccessedViaReflection`/`DynamicDependency` is fine too.